### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by dsergio**

### CODE REVIEW

**Comment by GamerBah** [comment]
Will review!

**Comment by GamerBah** [comment]
# Assignment 4

- WebAPI Analyzers enabled ✔
- DI properly configured
  - Services properly registered with correct scopes ✔
  - DB context properly registered ✔
  - AutoMapper properly registered ✔
- NSwag properly setup ✔
- API controllers properly expose CRUD operations ✔
- Controllers are all unit tested ✔

## Extra Credit
- Unit tests in EntityServiceTests fixed approprietly ✔
- Mocking framework used to create the test doubles ❌

I wasn't able to find where you used Moq, even though it's included in your packages and included in 2 of the controller test classes. Maybe I'm just looking over it, somehow?

File: SecretSanta/src/SecretSanta.Api/Controllers/EntityController.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using SecretSanta.Business;
+using SecretSanta.Data;
+
+namespace SecretSanta.Api.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class EntityController<TEntity> : ControllerBase where TEntity : class
```
Comment (GamerBah) [inline SecretSanta/src/SecretSanta.Api/Controllers/EntityController.cs]: Good job writing the controller generically for Entity, and extending that in the other controller classes.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+using System.Text;
+using SecretSanta.Data.Tests;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using System.Linq;
+
+namespace SecretSanta.Api.Tests.Controllers
+{
+	[TestClass]
+	public class GiftControllerTests
```
Comment (GamerBah) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs]: You can write generic test classes for all the controllers too, just like writing a generic controller class. [This is the site I found](http://www.codinghelmet.com/articles/how-to-write-unit-tests-for-generic-classes) very useful in explaining how to do it. Super handy for not having to write the same tests for different sub-classes.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UserControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+			}
+
+			public Task<User> InsertAsync(User entity)
+			{
+				int id = Items.Count + 1;
+				Items[id] = new TestUser(entity, id);
+				return Task.FromResult(Items[id]);
+			}
+
+			public Task<User[]> InsertAsync(params User[] entity)
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UserControllerTests.cs]: Parameter entity of method InsertAsync is never used. Remove the parameter or use it in the method body.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			}
+
+			public Task<Group> InsertAsync(Group entity)
+			{
+				int id = Items.Count + 1;
+				Items[id] = new TestGroup(entity, id);
+				return Task.FromResult(Items[id]);
+			}
+
+			public Task<Group[]> InsertAsync(params Group[] entity)
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Parameter entity of method InsertAsync is never used. Remove the parameter or use it in the method body.

File: SecretSanta/Directory.Build.props [hunk 5]
```diff
@@ -2,6 +2,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <Nullable>enable</Nullable>
+    <IncludeOpenAPIAnalyzers>true</IncludeOpenAPIAnalyzers>
```
Comment (Keboo) [inline SecretSanta/Directory.Build.props]: This certainly woks, but outside of web projects it does not add any value

File: SecretSanta/src/SecretSanta.Business/AutomapperConfigurationProfile.cs [hunk 6]
```diff
@@ -8,8 +8,12 @@ public class AutomapperConfigurationProfile : Profile
     {
         public AutomapperConfigurationProfile()
         {
-            CreateMap<Gift, Gift>().ForMember(property => property.Id, option => option.Ignore());
-            CreateMap<User, User>().ForMember(property => property.Id, option => option.Ignore());
+            CreateMap<Gift, Gift>()
+                .ForMember(property => property.Id, option => option.Ignore())
+                .ForMember(property => property.CreatedBy, option => option.Ignore());
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/AutomapperConfigurationProfile.cs]: Though this solution certainly works to make the unit tests happy, it does expose some problems. For example it still allows someone outside of the data layer to directly modify the finger printing properties (which really should be completely controlled in the data layer). You can see a solution in the Assignment4Complete branch.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs (hunk 2 above)
Comment (Keboo) [reply SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs]: Agreed, in general if there is a polymorphic relationship between the classes you are testing the same relationship can/should be applied to the tests.

File: SecretSanta/test/SecretSanta.Data.Tests/SecretSanta.Data.Tests.csproj [hunk 7]
```diff
@@ -14,6 +15,7 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="NSwag.AspNetCore" Version="13.2.2" />
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Data.Tests/SecretSanta.Data.Tests.csproj]: This should not need to get applied outside of the web projects.

**Pull request closed without merging**